platform/admin/admin-configuration-configinstaller/src/main/java/org/codice/ddf/admin/configuration/ConfigurationInstaller.java

```diff
@@ -0,0 +1,218 @@
+ * </tr>
+ * <tr>
+ * <td>Y</td><td>null</td><td>Felix file prop no longer being tracked, delete config file and remove from map</td>
+ * </tr>
+ * <tr>
+ * <td>Y</td><td>X</td><td>Felix file prop updated, delete config file and update map</td>
+ * </tr>
+ * <tr>
+ * <td>Y</td><td>Y</td><td>Felix file prop did not change, do nothing</td>
+ * </tr>
+ * </table>
+ *
+ * @see #handleUpdate(ConfigurationEvent)
+ * @see #handleDelete(ConfigurationEvent)
+ */
+public class ConfigurationInstaller implements SynchronousConfigurationListener {
+    private static final Logger LOGGER = LoggerFactory.getLogger(ConfigurationInstaller.class);
+
+    private final ConfigurationAdmin configurationAdmin;
+
+    private final ConfiguratorFactory configuratorFactory;
+
+    private final ConfigActions configActions;
+
+    private Map<String, File> pidFileMap;
+
+    public ConfigurationInstaller(ConfigurationAdmin configurationAdmin,
+            ConfiguratorFactory configuratorFactory, ConfigActions configActions) {
+        this.configurationAdmin = configurationAdmin;
+        this.pidFileMap = new ConcurrentHashMap<>();
+
+        this.configuratorFactory = configuratorFactory;
+        this.configActions = configActions;
+    }
+
+    /**
+     * @return a read-only map of configuration pids to felix file install properties.
+     */
+    Map<String, File> getPidFileMap() {
+        return Collections.unmodifiableMap(pidFileMap);
+    }
+
+    public void init() throws IOException, InvalidSyntaxException {
+        Configuration[] configs = configurationAdmin.listConfigurations(null);
+        if (configs == null) {
+            return;
+        }
+        this.pidFileMap = Arrays.stream(configs)
+                .map(FelixConfig::new)
+                .filter(config -> config.getFile() != null)
+                .collect(Collectors.toConcurrentMap(FelixConfig::getPid, FelixConfig::getFile,
+                        // API guarantees no duplicates in production, but itests violate this
+                        // Workaround is to shut off admin-configuration-passwordencryption
+                        (pid1, pid2) -> pid2));
+    }
+
+    @Override
+    public void configurationEvent(ConfigurationEvent event) {
+        String pid = event.getPid();
+        switch (event.getType()) {
+        case CM_UPDATED:
+            LOGGER.debug("Handling update for pid {}", pid);
+            handleUpdate(event);
+            break;
+        case CM_DELETED:
+            LOGGER.debug("Handling delete for pid {}", pid);
+            handleDelete(event);
+            break;
+        default:
+            LOGGER.debug("Unknown configuration event type, taking no action for pid {}", pid);
+        }
+    }
+
+    /**
+     * Sync the config files in etc when an update occurs.
+     *
+     * @param event the configuration event for which an update is needed
+     */
+    private void handleUpdate(ConfigurationEvent event) {
```

Comment:
✏️ I'm usually loath to suggest over-commenting code inline, but I think in this case it might be good. You could update the javadoc, but I think inline would be even clearer to future maintainers.

Unless I'm mistaken, you've five distinct operations that can occur, right?
1. No old file + no new file
2. No old file + new file
3. Old file + no new file
4. Old file + unchanged file
5. Old file + changed file

As a developer, in a method with this many logical branches, I would personally like to see inline comments indicating which branch is being addressed. It's the kind of thing I can certainly work out by reading the code and walking the conditions, but it's nice not to have to do so.